Spring Boot single-page app tutorial (#883)
oktadeveloper/okta-blog: _source/_posts/2021-10-04-spring-boot-spa.md

---
layout: blog_post
title: "Learn How to Build a Single-Page App with Vue and Spring Boot"
author: andrew-hughes
by: contractor
communities: [java,javascript]
description: ""
tags: [vue, java, spring-boot, spring, oauth2, oidc, spa, javascript]

> tweets:

Reviewer: ```suggestion
tweets:
"Use Spring Boot to create a simple resource server and Vue to create a front-end client. Step-by-step to a secure SPA with Spring Boot, Vue and @Okta. With some CORS for starters."
```

> - ""

Reviewer: ```suggestion
- "Everything you need to know to create a secure single-page application (SPA) with a Spring Boot resource server, a Vue front-end client, OAuth 2.0, OIDC & the @Okta Sign-In Widget "
```

- ""
- ""
image: blog/spring-boot-spa/spring-boot-spa-social.jpg
type: conversion
---

In this tutorial, you are going to create a single-page application (SPA) that uses a Spring Boot resource server and a Vue front-end client. You'll see how to configure Spring Boot to use JSON Web Tokens (JWT) for authentication and authorization, with Okta as an OAuth 2. 0 and OpenID Connect (OIDC) provider. You'll also see how to bootstrap a Vue client app with the Vue CLI and how to secure it using the Okta Sign-In Widget.

> Okta is a computer security services company that provides a lot of great resources for securing web applications. The widget is a great way to secure front-end applications because it allows you to easily add a secure login form that is configurable for single sign-on and social sign-on with external providers such as Google, Facebook, and LinkedIn. It provides an implementation of the authorization code OAuth 2.0 flow using PKCE (Proof Key for Code Exchange).

Reviewer 2: ```suggestion
Okta is a computer security services company that provides a lot of great resources for securing web applications. The Okta Sign-In Wdiget is a great way to secure front-end applications because it allows you to easily add a secure login form that is configurable for single sign-on and social sign-on with external providers such as Google, Facebook, and LinkedIn. It provides an implementation of the authorization code OAuth 2.0 flow using PKCE (Proof Key for Code Exchange).
```

---
layout: blog_post
title: "Learn How to Build a Single-Page App with Vue and Spring Boot"
author: andrew-hughes
by: contractor
communities: [java,javascript]

> description: ""

Reviewer 2: ```suggestion
description: "Tutorial: Build a secure application with a Vue.js frontend and a Spring Boot backend."
```

tags: [vue, java, spring-boot, spring, oauth2, oidc, spa, javascript]
tweets:
"Use Spring Boot to create a simple resource server and Vue to create a front-end client. Step-by-step to a secure SPA with Spring Boot, Vue and @Okta. With some CORS for starters."
- "Everything you need to know to create a secure single-page application (SPA) with a Spring Boot resource server, a Vue front-end client, OAuth 2.0, OIDC & the @Okta Sign-In Widget "

> - ""
> - ""

Reviewer 2: ```suggestion
```

**Table of Contents**{: .hide }
* Table of Contents
{:toc}

**Requirements**

Before you get started, you'll need to make sure you have a few tools installed.

- **Okta CLI**: The Okta CLI is an easy way to create projects that use Okta for security. Follow the installation instructions on the [Okta CLI project website](https://cli.okta.com/manual/#installation). You should log in to your existing account or register for a new account using the CLI before continuing with this tutorial.

- **Java 11**: This project uses Java 11. OpenJDK 11 will work just as well.  Instructions are found on the [OpenJDK website](https://openjdk.java.net/install/). OpenJDK can also be installed using [Homebrew](https://brew.sh/). Alternatively, [SDKMAN](https://sdkman.io/) is another excellent option for installing and managing Java versions.

- **Vue CLI**: You'll use the Vue CLI to generate the starter Vue client project. Install it according to the [the instructions on the Vue website](https://cli.vuejs.org/guide/installation.html).

> ## Create The Spring Boot Resource Server

Reviewer: ```suggestion
## Create the Spring Boot Resource Server
```
consistent caps

- Browser sends the origin request to the server
- Server handles original request, authenticating and authorizing it using OAuth 2.0 and OIDC before returning a reply

From a developer perspective, CORS is something that needs to be configured on the server. Spring Boot makes this easy. There are two places in the Java code where CORS is configured. CORS must generally be enabled for the Spring Boot application. This is done in the `SecurityConfiguration` class in the `configure(HttpSecurity http)` method by adding the `cors()` method to the `http` configuration chain.

The second place is the `@CrossOrigin` annotation on the `CaffeineLevelRestController` class. This tells Spring Boot to configure CORS for all endpoints in this controller. By default, this will allow cross-origin requests from any domain. To restrict requests to only our client application domain, we could have used the annotation `@CrossOrigin(origins = "http://localhost:8080")`.

Take a look at [this Spring blog post](https://spring.io/blog/2015/06/08/cors-support-in-spring-framework) for more info on Spring Boot and CORS.

> ## Configure The Resource Server For Okta Auth

Reviewer: ```suggestion
## Configure the Resource Server for Okta Auth
```
capitalization

{% img blog/spring-boot-spa/auth-servers.png alt:"Okta Dashboard Authorization Servers screenshot" width:"800" %}{: .center-image }

Add the issuer URI to your properties file, which should look something like below, replacing `{yourOktaDomain}` with your actual Okta domain.

`src/main/resources/application.properties`

```properties
server.port=8082
okta.oauth2.issuer=https://{yourOktaDomain}/oauth2/default
```

You now have a working, secure resource server. Go ahead and run the application.

```bash
./mvnw spring-boot:run
```

You should get a lot of console output that ends with the following.

```bash
2021-09-01 11:49:42.136  INFO 370077 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8082 (http) with context path ''
2021-09-01 11:49:42.143  INFO 370077 --- [           main] com.okta.springbootspa.DemoApplication   : Started DemoApplication in 2.021 seconds (JVM running for 2.199)
```

> ## Use The Vue CLI To Bootstrap The Client App

Reviewer: ```suggestion
## Use the Vue CLI to Bootstrap the Client App
```

```bash
npm install --save \
@okta/okta-signin-widget \
@okta/okta-vue \
@okta/okta-auth-js \
axios
```

Create a `src/okta/index.js` file:

```js
import OktaSignIn from '@okta/okta-signin-widget'
import { OktaAuth } from '@okta/okta-auth-js'

const yourOktaUri = '${yourOktaUri}';
const clientId = '${clientId}';

const oktaSignIn = new OktaSignIn({
    baseUrl: yourOktaUri,
    clientId: clientId,
    redirectUri: 'http://localhost:8080/login/callback',
    authParams: {
        pkce: true,
        issuer: `${yourOktaUri}/oauth2/default`,
        display: 'page',
        scopes: ['openid', 'profile', 'email']
    }
});

const oktaAuth = new OktaAuth({
    issuer: `${yourOktaUri}/oauth2/default`,
    clientId: clientId,
    redirectUri: window.location.origin + '/login/callback',
    scopes: ['openid', 'profile', 'email']
})

export { oktaAuth, oktaSignIn };
```

In the file above, **you need to replace the placeholders `${...}` with the appropriate values**. Before you can do this, however, you need to use the CLI to create the OIDC application on the Okta servers, which you will do directly below.

> ## Create A Single-Page OIDC Application On Okta

Reviewer: ```suggestion
## Create a Single-Page OIDC Application on Okta
```

This time you're going to create a single-page application for the front-end (instead of the web application type you used for the resource server). Enter the following command into a shell.

```bash
okta apps create
```

Press `enter` to accept the default **Application name** `okta-app`. (Application name here refers to the OIDC application created on the Okta servers).

For **Type of Application**, Select `2: Single Page App`.

Change the **Redirect URI** to `http://localhost:8080/login/callback`.

Press `enter` to accept the default **Post Logout Redirect**.

You should see some output like the following.

```bash
Okta application configuration: 
Issuer:    https://{yourOktaUri}/oauth2/default
Client ID: {clientId}
```

The values in brackets will, of course, be your actual values.

**You need to go back to the `src/okta/index.js` file and replace the `yourOktaUri` and `clientId` variables at the top of the file with your values.** The value for `yourOktaUri` is the Okta URI without any further path specifiers. It will look like this: `https://dev-133337.okta.com`

> One of the things that the Okta CLI does for you is to add your application's base URL to the CORS trusted origins for the Okta auth server. This is necessary because the the Okta Sign-In Widget will be making cross0-origin requests, that, as noted above, would be blocked unless CORS is properly handled. You can see this by going to your Okta developer dashboard and selected **Security** and **API** from the left-side menu and selected **Trusted Origins** from the API tab. You'll see that (in this case) `http://localhost:8080` was added as a trusted origin of type `CORS Redirect`.

Reviewer 2: ```suggestion
One of the things that the Okta CLI does for you is to add your application's base URL to the CORS trusted origins for the Okta auth server. This is necessary because the the Okta Sign-In Widget will be making cross-origin requests, that, as noted above, would be blocked unless CORS is properly handled. You can see this by going to your Okta developer dashboard. with **Security** and **API** selected from the left-side menu, and **Trusted Origins** selected from the API tab. You'll see that (in this case) `http://localhost:8080` was added as a trusted origin of type `CORS Redirect`.
```

Reviewer 2: Adding back in @havi's suggestion

{% img blog/spring-boot-spa/trusted-origin.png alt:"Trusted origins for Okta CORS" width:"800" %}{: .center-image }

You can read more about CORS and Okta in the [Okta documentation](https://developer.okta.com/docs/guides/enable-cors/overview/).

> ## Finish the Vue Front-End Client App

Reviewer: ```suggestion
## Finish The Vue Front-End Client App
```

> ## Finish The Vue Front-End Client App

Reviewer: ```suggestion
## Finish the Vue Front-End Client App
```

> The next file you need to create is a wrapper for the Okta Sign-In widget.

Reviewer: ```suggestion
The next file you need to create is a wrapper for the Okta Sign-In Widget.
```
consistent capitalization
